pyclam/criterion.py

```diff
@@ -127,131 +129,45 @@ def __call__(self, root: Cluster) -> Set[Cluster]:
+    """ Uses a decision function from a trained meta-ml model to select best clusters for graph.
     """
-    # TODO: Change this to use a parsed function, just like Regression Tree Criterion.
-
-    def __init__(self, constants: Union[np.array, List[float]], *, mode: str = 'ranked'):
-        """ A Linear Regression based Meta-ML model to select Clusters for a Graph.
+    def __init__(self, predict_auc: Callable[[np.array], float], min_depth: int = 4):
+        """ A Meta-ML model to select Clusters for a Graph.
 
-        The constants correspond to:
-            * child/parent lfd ratio
-            * child/parent cardinality ratio
-            * child/parent radii ratio
-            * lfd-ratio exponential-moving-average (ema)
-            * cardinality-ratio ema
-            * radii-ratio ema
-
-        :param constants: 6 constants from the LR model.
-        :param mode: which selection method to use. Must be one of 'ranked' or 'percentile'.
+        :param predict_auc: A function that takes a cluster's ratios and returns a predicted auc for selecting that cluster.
         """
-        if mode not in SELECTION_MODES.keys():
-            raise ValueError(f'mode must be one of {list(SELECTION_MODES.keys())}. Gor {mode} instead.')
-        self.mode = SELECTION_MODES[mode]
-
-        num_constants = 6
-        constants = np.asarray(constants, dtype=float)
-        if constants.shape != (num_constants,):
-            raise ValueError(f'expected a vector of {num_constants} elements. Got {constants.shape} instead.')
-        self.constants: np.array = constants
+        self.predict_auc: Callable[[np.array], float] = predict_auc
+        if min_depth < 1:
+            raise ValueError(f'min-depth must be a positive integer.')
+        self.min_depth = min_depth
 
     def __call__(self, root: Cluster) -> Set[Cluster]:
-        predicted_auc: Dict[Cluster, float] = self._predict_auc(root.manifold)
-        return self.mode(root.manifold, predicted_auc)
+        def predict_auc(cluster):
+            # adjust predicted auc by cardinality to slightly favor clusters with high cardinalities.
+            return self.predict_auc(manifold.cluster_ratios(cluster)) * (1 + cluster.cardinality / root.cardinality / 10)
 
-    def _predict_auc(self, manifold: Manifold) -> Dict[Cluster, float]:
-        return {
-            cluster: float(np.dot(self.constants, manifold.cluster_ratios(cluster)))
+        manifold: Manifold = root.manifold
+        predicted_auc: Dict[Cluster, float] = {
+            cluster: predict_auc(cluster)
             for cluster in manifold.ordered_clusters
+            if cluster.depth >= self.min_depth
         }
+        normalized_auc = normalize(np.asarray([-v for v in predicted_auc.values()]), mode='gaussian')
+        # Turn dict into max priority queue
+        items = list(zip(normalized_auc, predicted_auc.keys()))
+        heapq.heapify(items)
 
-
-class RegressionTreePaths(SelectionCriterion):
```

Comment:
Why remove the existing selection criteria?

Reply:
They are redundant. The new criterion is better